### PR TITLE
Update only relevant cols on repo sync

### LINKED ISSUE #511: RepoSync disable enabled repos
#510 did fix a bug but create a new one ...

now let's do it properly!!!

### PR DESCRIPTION
**Pull request by 6543**
close  #511

### CODE REVIEW

**Comment by 6543** [comment]
now unit tests cover this bug: https://wp.laszlo.cloud/woodpecker-ci/woodpecker/1160/17

File: server/store/datastore/permission.go [hunk 1]
```diff
@@ -29,19 +31,26 @@ func (s storage) PermFind(user *model.User, repo *model.Repo) (*model.Perm, erro
...
-	// TODO: check if needed
-	if len(perm.Repo) != 0 {
+	if err := s.permUpsert(sess, perm); err != nil {
+		return err
+	}
+
+	return sess.Commit()
+}
+
+func (s storage) permUpsert(sess *xorm.Session, perm *model.Perm) error {
```
Comment (anbraten) [inline server/store/datastore/permission.go:47]: What was the reason for creating a new function here?

File: server/shared/userSyncer.go [hunk 2]
```diff
@@ -70,24 +71,24 @@ func (s *Syncer) Sync(ctx context.Context, user *model.User) error {
...
 			}
 			remotePerm, err := s.Remote.Perm(ctx, user, repo.Owner, repo.Name)
-			if err == nil && remotePerm != nil {
-				perm.Push = remotePerm.Push
-				perm.Admin = remotePerm.Admin
+			if err != nil {
+				return fmt.Errorf("could not fetch permission of repo '%s': %v", repo.FullName, err)
 			}
-			perms = append(perms, &perm)
+			repo.Perm.Pull = remotePerm.Pull
```
Comment (anbraten) [inline server/shared/userSyncer.go:87]: So by this change we wont save permissions on user-sync to the database? Wasn't the `s.Remote.Perm(ctx, user, repo.Owner, repo.Name)` call the expensive part?

File: server/store/datastore/repo.go [hunk 3]
```diff
@@ -86,25 +88,42 @@ func (s storage) RepoBatch(repos []*model.Repo) error {
...
+				return err
+			}
 		} else {
 			// only Insert on single object ref set auto created ID back to object
 			if _, err := sess.Insert(repos[i]); err != nil {
 				return err
 			}
 		}
+
+		if repos[i].Perm != nil {
```
Comment (anbraten) [inline server/store/datastore/repo.go:120]: This means: If we do get permissions for the current users for this repo back, we save those permissions, right?

**Review by anbraten — COMMENTED** [review]
Sorry for my questions. I am still not 100% sure how permissions will be saved and updated. Maybe we should think about removing `repo.Perm` as those are not repo permissions, but the permissions a specific user has for this repo.

File: server/store/datastore/permission.go (hunk 1 above)
Comment (6543) [reply server/store/datastore/permission.go:47]: to have it within one sql transaction (**`sess *xorm.Session`**) - it's a private function so it's fine to do so :)

File: server/store/datastore/repo.go (hunk 3 above)
Comment (6543) [reply server/store/datastore/repo.go:120]: yes and that's what syncer now does

File: server/shared/userSyncer.go (hunk 2 above)
Comment (6543) [reply server/shared/userSyncer.go:87]: this is no fix to the other issue :/ - it make it easyer to fix sure but that's not the scope of that pull ... (it already got bit out of scope :/ )

File: server/store/datastore/repo.go [hunk 4]
```diff
@@ -74,6 +74,8 @@ func (s storage) RepoList(user *model.User, owned bool) ([]*model.Repo, error) {
 		Find(&repos)
 }
 
+// RepoBatch Sync batch of repos (with permissions) to store (create if not exist else update)
```
Comment (anbraten) [inline server/store/datastore/repo.go:77]:
```suggestion
// RepoBatch Sync batch of repos from SCM (with permissions) to store (create if not exist else update)
```

**Review by anbraten — APPROVED** [review]

**Pull request merged**